Index: app/src/main/java/mvasoft/timetracker/core/TimeTrackerApp.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- app/src/main/java/mvasoft/timetracker/core/TimeTrackerApp.java	(revision )
+++ app/src/main/java/mvasoft/timetracker/core/TimeTrackerApp.java	(revision )
@@ -0,0 +1,13 @@
+package mvasoft.timetracker.core;
+
+import dagger.android.AndroidInjector;
+import dagger.android.support.DaggerApplication;
+
+
+public class TimeTrackerApp extends DaggerApplication {
+
+    @Override
+    protected AndroidInjector<? extends DaggerApplication> applicationInjector() {
+        return DaggerAppComponent.builder().create(this);
+    }
+}
Index: app/src/main/java/mvasoft/timetracker/BackupAssistant.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- app/src/main/java/mvasoft/timetracker/BackupAssistant.java	(date 1510424560000)
+++ app/src/main/java/mvasoft/timetracker/BackupAssistant.java	(revision )
@@ -16,6 +16,7 @@
 public class BackupAssistant {
 
     private static final String BACKUP_FOLDER_NAME = "TimeTrackerBackups";
+
     private final Context mContext;
 
     public BackupAssistant(Context context) {
Index: build.gradle
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- build.gradle	(date 1510424560000)
+++ build.gradle	(revision )
@@ -22,6 +22,10 @@
     }
 }
 
+ext {
+    dagger = '2.13'
+}
+
 task clean(type: Delete) {
     delete rootProject.buildDir
 }
Index: app/build.gradle
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- app/build.gradle	(date 1510424560000)
+++ app/build.gradle	(revision )
@@ -28,6 +28,12 @@
 }
 
 dependencies {
+    compile "com.google.dagger:dagger:$rootProject.ext.dagger"
+    compile "com.google.dagger:dagger-android:$rootProject.ext.dagger"
+    compile "com.google.dagger:dagger-android-support:$rootProject.ext.dagger"
+    annotationProcessor "com.google.dagger:dagger-compiler:$rootProject.ext.dagger"
+    annotationProcessor "com.google.dagger:dagger-android-processor:$rootProject.ext.dagger"
+
     compile fileTree(include: ['*.jar'], dir: 'libs')
     androidTestCompile('com.android.support.test.espresso:espresso-core:2.2.2', {
         exclude group: 'com.android.support', module: 'support-annotations'
@@ -41,9 +47,4 @@
     compile project(':app:libs:twoway-view_core')
     compile project(':app:libs:twoway-view_layouts')
     testCompile 'junit:junit:4.12'
-    compile 'com.google.dagger:dagger:2.12'
-    compile 'com.google.dagger:dagger-android-support:2.12'
-    compile 'com.google.dagger:dagger-android-processor:2.12'
-    annotationProcessor 'com.google.dagger:dagger-compiler:2.12'
-    annotationProcessor 'com.google.dagger:dagger-android-processor:2.12'
 }
Index: app/src/main/java/mvasoft/timetracker/data/SessionsContentProvider.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- app/src/main/java/mvasoft/timetracker/data/SessionsContentProvider.java	(date 1510424560000)
+++ app/src/main/java/mvasoft/timetracker/data/SessionsContentProvider.java	(revision )
@@ -3,7 +3,7 @@
 import android.content.ContentProvider;
 import android.content.ContentResolver;
 import android.content.ContentValues;
-import android.content.Intent;
+import android.content.Context;
 import android.content.UriMatcher;
 import android.database.Cursor;
 import android.database.SQLException;
@@ -14,12 +14,12 @@
 import javax.inject.Inject;
 
 import dagger.Lazy;
-import mvasoft.timetracker.SessionsWidgetService;
-import mvasoft.timetracker.core.WidgetHelper;
+import dagger.android.DaggerContentProvider;
 import mvasoft.timetracker.data.DatabaseDescription.GroupsDescription;
 import mvasoft.timetracker.data.DatabaseDescription.SessionDescription;
+import mvasoft.timetracker.widget.IWidgetHelper;
 
-public class SessionsContentProvider extends ContentProvider {
+public class SessionsContentProvider extends DaggerContentProvider {
 
     private static final UriMatcher mMatcher = new UriMatcher(UriMatcher.NO_MATCH);
     private static final int ONE_SESSION        = 1;
@@ -85,14 +85,16 @@
     }
 
     @Inject
-    Lazy<WidgetHelper> mWidgetHelper;
+    Lazy<IWidgetHelper> mWidgetHelper;
     private SessionsDatabaseHelper mDBHelper;
 
     public SessionsContentProvider() {
+        super();
     }
 
     @Override
     public boolean onCreate() {
+        super.onCreate();
         mDBHelper = new SessionsDatabaseHelper(getContext());
         return true;
     }
Index: app/src/main/java/mvasoft/timetracker/core/AppComponent.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- app/src/main/java/mvasoft/timetracker/core/AppComponent.java	(revision )
+++ app/src/main/java/mvasoft/timetracker/core/AppComponent.java	(revision )
@@ -0,0 +1,21 @@
+package mvasoft.timetracker.core;
+
+import javax.inject.Singleton;
+
+import dagger.Component;
+import dagger.android.AndroidInjector;
+import dagger.android.support.AndroidSupportInjectionModule;
+import mvasoft.timetracker.BackupAssistant;
+import mvasoft.timetracker.widget.WidgetModule;
+
+@Singleton
+@Component(modules = {AndroidSupportInjectionModule.class,
+        AppModule.class,
+        WidgetModule.class})
+public interface AppComponent extends AndroidInjector<TimeTrackerApp> {
+
+    void inject(BackupAssistant assistant);
+
+    @Component.Builder
+    abstract class Builder extends AndroidInjector.Builder<TimeTrackerApp> {}
+}
Index: app/src/main/java/mvasoft/timetracker/core/AppModule.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- app/src/main/java/mvasoft/timetracker/core/AppModule.java	(revision )
+++ app/src/main/java/mvasoft/timetracker/core/AppModule.java	(revision )
@@ -0,0 +1,19 @@
+package mvasoft.timetracker.core;
+
+import android.content.Context;
+
+import dagger.Binds;
+import dagger.Module;
+import dagger.android.ContributesAndroidInjector;
+import mvasoft.timetracker.data.SessionsContentProvider;
+import mvasoft.timetracker.widget.WidgetModule;
+
+@Module
+public abstract class AppModule {
+
+    @Binds
+    abstract Context bindApplicationContext(TimeTrackerApp app);
+
+    @ContributesAndroidInjector(modules = {WidgetModule.class})
+    abstract SessionsContentProvider sessionsContentProviderInjector();
+}
Index: app/src/main/java/mvasoft/timetracker/widget/WidgetModule.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- app/src/main/java/mvasoft/timetracker/widget/WidgetModule.java	(revision )
+++ app/src/main/java/mvasoft/timetracker/widget/WidgetModule.java	(revision )
@@ -0,0 +1,14 @@
+package mvasoft.timetracker.widget;
+
+import javax.inject.Singleton;
+
+import dagger.Binds;
+import dagger.Module;
+
+@Module
+public abstract class WidgetModule {
+
+    @Binds
+    @Singleton
+    abstract IWidgetHelper bindWidgetHelper(WidgetHelper impl);
+}
Index: app/src/main/java/mvasoft/timetracker/widget/IWidgetHelper.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- app/src/main/java/mvasoft/timetracker/widget/IWidgetHelper.java	(revision )
+++ app/src/main/java/mvasoft/timetracker/widget/IWidgetHelper.java	(revision )
@@ -0,0 +1,5 @@
+package mvasoft.timetracker.widget;
+
+public interface IWidgetHelper {
+    void updateWidget();
+}
Index: app/src/main/java/mvasoft/timetracker/widget/WidgetHelper.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- app/src/main/java/mvasoft/timetracker/widget/WidgetHelper.java	(revision )
+++ app/src/main/java/mvasoft/timetracker/widget/WidgetHelper.java	(revision )
@@ -0,0 +1,28 @@
+package mvasoft.timetracker.widget;
+
+import android.content.Context;
+import android.content.Intent;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import mvasoft.timetracker.SessionsWidgetService;
+
+@Singleton
+public class WidgetHelper implements IWidgetHelper {
+
+    private Context mContext;
+
+    @Inject
+    WidgetHelper(Context context) {
+        super();
+        mContext = context;
+    }
+
+    @Override
+    public void updateWidget() {
+        Intent serviceIntent = new Intent(mContext, SessionsWidgetService.class);
+        serviceIntent.setAction(SessionsWidgetService.ACTION_UPDATE_WIDGET);
+        mContext.startService(serviceIntent);
+    }
+}
